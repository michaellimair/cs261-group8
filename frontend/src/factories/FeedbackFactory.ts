/* eslint-disable class-methods-use-this */
import faker from '@faker-js/faker';
import { FeedbackType, IFeedback } from 'customTypes/feedback';
import { merge } from 'lodash';
import FeedbackReplyFactory from './FeedbackReplyFactory';

class FeedbackFactory {
  create = (feedback?: Partial<IFeedback>): IFeedback => {
    const replyFactory = new FeedbackReplyFactory();
    return merge({
      id: faker.datatype.number(),
      title: faker.lorem.lines(1),
      content: faker.lorem.paragraph(3),
      type: FeedbackType.BUG,
      created: new Date(),
      modified: new Date(),
      reply: replyFactory.create(),
    }, feedback);
  };

  createMany = (count: number) => Array(count).fill(null).map((_, i) => this.create({ id: i }));
}

export default FeedbackFactory;
