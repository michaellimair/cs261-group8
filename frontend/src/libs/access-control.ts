import { IUserGroup, UserGroup } from 'customTypes/auth';

/** Function to check whether a user is allowed to access a certain resource
 * given the list of groups which can access the resource. */
export const checkAllowed = ({
  userGroups,
  allowedGroups,
}: {
  userGroups: IUserGroup[];
  allowedGroups: UserGroup[];
}): boolean => {
  if (allowedGroups.includes(UserGroup.UNDECLARED) && !userGroups.length) {
    return true;
  }
  return allowedGroups
    .some((groupName) => userGroups.find(({ name }) => name === groupName));
};
