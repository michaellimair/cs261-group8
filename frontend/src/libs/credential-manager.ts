export const JWT_TOKEN_KEY = 'jwt_token';
export const JWT_EXPIRY_KEY = 'jwt_expiry';

class CredentialManager {
  protected token: string | null = null;

  protected expiry: Date | null = null;

  constructor(private readonly storage: Storage = localStorage) {
    this.restoreAuthenticationCredentials();
  }

  private restoreAuthenticationCredentials() {
    this.token = this.storage.getItem(JWT_TOKEN_KEY);
    const expiryStr = this.storage.getItem(JWT_EXPIRY_KEY);

    if (expiryStr) {
      this.expiry = new Date(expiryStr);
    }

    if (this.isCredentialExpired()) {
      this.clearCredentials();
    }
  }

  private isCredentialExpired(): boolean {
    return this.expiry !== null && (new Date()).getTime() > this.expiry.getTime();
  }

  setCredentials(token: string, expiry: Date) {
    this.token = token;
    this.expiry = expiry;
    this.storage.setItem(JWT_TOKEN_KEY, this.token);
    this.storage.setItem(JWT_EXPIRY_KEY, this.expiry.toISOString());
  }

  clearCredentials() {
    this.storage.removeItem(JWT_TOKEN_KEY);
    this.storage.removeItem(JWT_EXPIRY_KEY);

    this.token = null;
    this.expiry = null;
  }

  get credentials() {
    if (this.isCredentialExpired()) {
      this.clearCredentials();
    }
    return {
      token: this.token,
      expiry: this.expiry,
    };
  }
}

export default CredentialManager;
