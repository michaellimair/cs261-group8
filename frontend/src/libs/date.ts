export const formatDateTime = (date: Date, timeZone?: string): string => new Intl.DateTimeFormat('en-GB', {
  weekday: 'long',
  timeZoneName: 'short',
  day: 'numeric',
  month: 'long',
  year: 'numeric',
  hour: 'numeric',
  minute: 'numeric',
  timeZone,
}).format(date);

export const formatTime = (date: Date, timeZone?: string): string => new Intl.DateTimeFormat('en-GB', { hour: 'numeric', minute: 'numeric', timeZone }).format(date);

export const formatDate = (date: Date, timeZone?: string): string => new Intl.DateTimeFormat('en-GB', {
  day: 'numeric', month: 'short', year: 'numeric', timeZone,
}).format(date);
