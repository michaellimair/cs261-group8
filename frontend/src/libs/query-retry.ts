import NotFoundError from 'api/error/NotFoundError';
import UnauthorizedError from 'api/error/UnauthorizedError';
import { RetryValue } from 'react-query/types/core/retryer';

export const queryRetryCondition: Exclude<RetryValue<unknown>, number | boolean> = (
  retryCount,
  err,
) => retryCount < 3
    && !NotFoundError.isNotFoundError(err)
    && !UnauthorizedError.isUnauthorizedError(err);
