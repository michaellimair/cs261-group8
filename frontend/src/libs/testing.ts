export const tick = (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});
