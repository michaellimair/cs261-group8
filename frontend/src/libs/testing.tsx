import { FC } from 'react';

import {
  MemoryRouter, Route, Routes,
} from 'react-router-dom';

export const tick = (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});

export const QueryRouterWrapper: FC = ({ children }) => (
  <MemoryRouter initialEntries={['/']}>
    <Routes>
      <Route path="*" element={children} />
    </Routes>
  </MemoryRouter>
);
