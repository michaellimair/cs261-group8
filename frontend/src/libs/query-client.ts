import { QueryClient } from 'react-query';
import { queryRetryCondition } from 'libs/query-retry';

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      // If error is not a 404 error, allow retry
      retry: queryRetryCondition,
      cacheTime: 0,
    },
  },
});
