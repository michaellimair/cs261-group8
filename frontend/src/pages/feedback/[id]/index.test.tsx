import { render } from '@testing-library/react';
import FeedbackFactory from 'factories/FeedbackFactory';
import { QueryRouterWrapper } from 'libs/testing';
import { useQuery } from 'react-query';
import ViewFeedbackPage from './index';

jest.mock('react-query', () => ({
  useQuery: jest.fn(),
  useMutation: jest.fn(),
}));

describe('pages/feedback/[id]', () => {
  it('returns data as usual when there are feedbacks', () => {
    const feedbackFactory = new FeedbackFactory();
    const data = feedbackFactory.createMany(10);
    (useQuery as jest.Mock).mockImplementationOnce(() => ({
      isLoading: false,
      data,
    }));

    const result = render(
      <QueryRouterWrapper>
        <ViewFeedbackPage />
      </QueryRouterWrapper>,
    );

    expect(result).toMatchSnapshot();
  });
});
