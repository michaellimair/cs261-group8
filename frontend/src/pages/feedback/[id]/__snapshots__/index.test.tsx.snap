// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/feedback/[id] returns data as usual when there are feedbacks 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="chakra-stack css-n21gh5"
      >
        <div
          class="chakra-stack css-buj9s7"
        >
          <a
            class="chakra-link css-wlea3r"
            href="/feedbacks"
          >
            &lt; dashboard.feedback.back_to_all
          </a>
          <h2
            class="chakra-heading css-0"
          >
            dashboard.feedback.title
          </h2>
        </div>
        <div
          class="css-jhpxgx"
        >
          <div
            class="chakra-stack css-84zodg"
          >
            <p
              class="chakra-text css-1d96wps"
            />
            <span
              class="css-1qm7lvz"
            >
              <svg
                class="chakra-icon css-1azbes1"
                fill="currentColor"
                focusable="false"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12,22 C17.5228475,22 22,17.5228475 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,17.5228475 6.4771525,22 12,22 Z M12,15 L12,14 C12,13 12,12.5 13,12 C14,11.5 15,11 15,9.5 C15,8.5 14,7 12,7 C10,7 9,8.26413718 9,10 M12,16 L12,18"
                  fill="none"
                  stroke="#000"
                  stroke-width="2"
                />
              </svg>
              <span
                class="css-8uhtka"
              >
                dashboard.feedback.undefined
              </span>
            </span>
          </div>
          <div
            class="chakra-stack css-1xqdd13"
          >
            <p
              class="chakra-text css-1k6unk6"
            />
            <p
              class="chakra-text css-1859sd4"
            >
              created
            </p>
            <p
              class="chakra-text css-1859sd4"
            >
              modified
            </p>
          </div>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-1upb9tn"
          />
          <div
            class="chakra-stack css-1x5vwo5"
          >
            <p
              class="chakra-text css-sfhhng"
            >
              dashboard.feedback.admin_reply
            </p>
            <p
              class="chakra-text css-1lj023j"
            >
              -
            </p>
          </div>
          <hr
            aria-orientation="horizontal"
            class="chakra-divider css-1upb9tn"
          />
          <div
            class="chakra-stack css-17vvmxc"
          >
            <a
              class="chakra-link chakra-button css-ogxy0s"
              href="/edit"
            >
              edit
            </a>
            <button
              class="chakra-button css-5c2176"
              type="button"
            >
              delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-stack css-n21gh5"
    >
      <div
        class="chakra-stack css-buj9s7"
      >
        <a
          class="chakra-link css-wlea3r"
          href="/feedbacks"
        >
          &lt; dashboard.feedback.back_to_all
        </a>
        <h2
          class="chakra-heading css-0"
        >
          dashboard.feedback.title
        </h2>
      </div>
      <div
        class="css-jhpxgx"
      >
        <div
          class="chakra-stack css-84zodg"
        >
          <p
            class="chakra-text css-1d96wps"
          />
          <span
            class="css-1qm7lvz"
          >
            <svg
              class="chakra-icon css-1azbes1"
              fill="currentColor"
              focusable="false"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12,22 C17.5228475,22 22,17.5228475 22,12 C22,6.4771525 17.5228475,2 12,2 C6.4771525,2 2,6.4771525 2,12 C2,17.5228475 6.4771525,22 12,22 Z M12,15 L12,14 C12,13 12,12.5 13,12 C14,11.5 15,11 15,9.5 C15,8.5 14,7 12,7 C10,7 9,8.26413718 9,10 M12,16 L12,18"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
            <span
              class="css-8uhtka"
            >
              dashboard.feedback.undefined
            </span>
          </span>
        </div>
        <div
          class="chakra-stack css-1xqdd13"
        >
          <p
            class="chakra-text css-1k6unk6"
          />
          <p
            class="chakra-text css-1859sd4"
          >
            created
          </p>
          <p
            class="chakra-text css-1859sd4"
          >
            modified
          </p>
        </div>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-1upb9tn"
        />
        <div
          class="chakra-stack css-1x5vwo5"
        >
          <p
            class="chakra-text css-sfhhng"
          >
            dashboard.feedback.admin_reply
          </p>
          <p
            class="chakra-text css-1lj023j"
          >
            -
          </p>
        </div>
        <hr
          aria-orientation="horizontal"
          class="chakra-divider css-1upb9tn"
        />
        <div
          class="chakra-stack css-17vvmxc"
        >
          <a
            class="chakra-link chakra-button css-ogxy0s"
            href="/edit"
          >
            edit
          </a>
          <button
            class="chakra-button css-5c2176"
            type="button"
          >
            delete
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
